loki@0.1.7 %gcc@11.2.0 build fails: SmallObj.h:462:57: error: ISO C++17 does not allow dynamic exception specifications

### Steps to reproduce the issue

Reproducible using:
* `spack@develop` (b32cb5765c598e59f2f91d99807b4331f4eba569 from `Fri Aug 12 23:17:20 2022 -0700`)
* `ecpe4s/ubuntu22.04-runner-x86_64:2022-08-01`
* `%gcc@11.2.0`

Concrete spec: [loki.spec.json.txt](https://github.com/spack/spack/files/9332403/loki.spec.json.txt)

Concretization:
```
 -   loki@0.1.7%gcc@11.2.0+shared arch=linux-ubuntu22.04-x86_64
```

Install error:
```
$> spack install -f ./loki.spec.json
==> Installing loki-0.1.7-uoibvp4a7lmgqffnrrdilqtid7ymgg4x
==> No binary for loki-0.1.7-uoibvp4a7lmgqffnrrdilqtid7ymgg4x found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/07/07553754f6be2738559947db69b0718512665bf4a34015fa3a875b6eb1111198.tar.bz2
==> No patches needed for loki
==> loki: Executing phase: 'edit'
==> loki: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' '-C' 'src' 'build-shared'

6 errors found in build log:
     66     /usr/include/c++/11/bits/unique_ptr.h:57:28: note: declared here
     67        57 |   template<typename> class auto_ptr;
     68           |                            ^~~~~~~~
     69     In file included from ../include/loki/SmartPtr.h:33,
     70                      from SmartPtr.cpp:20:
     71     ../include/loki/SmallObj.h: At global scope:
  >> 72     ../include/loki/SmallObj.h:462:57: error: ISO C++17 does not allow dynamic exception specifications
     73       462 |         static void * operator new ( std::size_t size ) throw ( std::bad_alloc )
     74           |                                                         ^~~~~
     75     In file included from ../include/loki/SmallObj.h:23,
     76                      from SmallObj.cpp:19:
     77     ../include/loki/Singleton.h: In function 'void Loki::SetLongevity(T*, unsigned int, Destroyer)':
     78     ../include/loki/Singleton.h:158:14: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]

     ...

     101                     from StrongPtr.cpp:16:
     102    /usr/include/c++/11/bits/unique_ptr.h:57:28: note: declared here
     103       57 |   template<typename> class auto_ptr;
     104          |                            ^~~~~~~~
     105    In file included from SmallObj.cpp:19:
     106    ../include/loki/SmallObj.h: At global scope:
  >> 107    ../include/loki/SmallObj.h:462:57: error: ISO C++17 does not allow dynamic exception specifications
     108      462 |         static void * operator new ( std::size_t size ) throw ( std::bad_alloc )
     109          |                                                         ^~~~~
     110    In file included from ../include/loki/SmartPtr.h:33,
     111                     from ../include/loki/StrongPtr.h:20,
     112                     from StrongPtr.cpp:16:
     113    ../include/loki/SmallObj.h: At global scope:
  >> 114    ../include/loki/SmallObj.h:462:57: error: ISO C++17 does not allow dynamic exception specifications
     115      462 |         static void * operator new ( std::size_t size ) throw ( std::bad_alloc )
     116          |                                                         ^~~~~
  >> 117    make: *** [Makefile:86: SmartPtr.lo] Error 1
     118    make: *** Waiting for unfinished jobs....
     119    In file included from ../include/loki/OrderedStatic.h:22,
     120                     from OrderedStatic.cpp:16:
     121    ../include/loki/Singleton.h: In function 'void Loki::SetLongevity(T*, unsigned int, Destroyer)':
     122    ../include/loki/Singleton.h:158:14: warning: 'template<class> class std::auto_ptr' is deprecated: use 'std::unique_ptr' instead [-Wdeprecated-declarations]
     123      158 |         std::auto_ptr<LifetimeTracker>

     ...

     126                     from ../include/loki/Singleton.h:30,
     127                     from ../include/loki/OrderedStatic.h:22,
     128                     from OrderedStatic.cpp:16:
     129    /usr/include/c++/11/bits/unique_ptr.h:57:28: note: declared here
     130       57 |   template<typename> class auto_ptr;
     131          |                            ^~~~~~~~
  >> 132    make: *** [Makefile:86: SmallObj.lo] Error 1
  >> 133    make: *** [Makefile:86: StrongPtr.lo] Error 1
     134    make: Leaving directory '/tmp/root/spack-stage/spack-stage-loki-0.1.7-uoibvp4a7lmgqffnrrdilqtid7ymgg4x/spack-src/src'
```

### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.19.0.dev0 (b32cb5765c598e59f2f91d99807b4331f4eba569)
* **Python:** 3.10.4
* **Platform:** linux-ubuntu22.04-zen2
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/9332401/spack-build-out.txt)
[spack-build-env.txt](https://github.com/spack/spack/files/9332402/spack-build-env.txt)

FYI @wspear

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/loki/package.py
+++ b/var/spack/repos/builtin/packages/loki/package.py
@@ -19,11 +19,12 @@ class Loki(MakefilePackage):
     variant("shared", default=True, description="Build shared libraries")
 
     def flag_handler(self, name, flags):
-        iflags = []
         if name == "cxxflags":
             if self.spec.satisfies("%oneapi@2023.0.0:"):
-                iflags.append("-Wno-error=dynamic-exception-spec")
-        return (iflags, None, None)
+                flags.append("-Wno-error=dynamic-exception-spec")
+            if self.spec.satisfies("@0.1.7 %gcc@11:"):
+                flags.append("-std=c++14")
+        return (flags, None, None)
 
     def build(self, spec, prefix):
         if "+shared" in spec:

